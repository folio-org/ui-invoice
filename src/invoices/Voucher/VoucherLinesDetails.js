import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { findIndex } from 'lodash';

import {
  Row,
  Headline,
  MultiColumnList,
} from '@folio/stripes/components';
import { AmountWithCurrencyField } from '@folio/stripes-acq-components';

import {
  getTotalAmount,
  groupByExternalAccNumber,
} from './utils';

const visibleColumns = ['lineNumber', 'group', 'fundCode', 'externalAccountNumber', 'amount'];
const columnMapping = {
  lineNumber: <FormattedMessage id="ui-invoice.voucher.voucherLines.lineNumber" />,
  group: <FormattedMessage id="ui-invoice.voucher.voucherLines.group" />,
  fundCode: <FormattedMessage id="ui-invoice.voucher.voucherLines.fundCode" />,
  externalAccountNumber: <FormattedMessage id="ui-invoice.voucher.voucherLines.externalAccountNumber" />,
  amount: <FormattedMessage id="ui-invoice.voucher.voucherLines.amount" />,
};
const columnWidths = {
  lineNumber: '20%',
  group: '20%',
  fundCode: '20%',
  externalAccountNumber: '20%',
  amount: '20%',
};

const VoucherLinesDetails = ({ voucherLines, currency }) => {
  const groupedVoucherLines = groupByExternalAccNumber(voucherLines);

  return (
    Object.values(groupedVoucherLines).map(lines => {
      const externalAccountNumber = lines[0].externalAccountNumber;
      const resultsFormatter = {
        lineNumber: line => findIndex(lines, ['id', line.id]) + 1,
        fundCode: ({ fundDistributions }) => fundDistributions.map(({ code }) => code).join(', '),
        amount: line => (
          <AmountWithCurrencyField
            amount={line.amount}
            currency={currency}
          />
        ),
      };

      return (
        <Fragment key={externalAccountNumber}>
          <Headline>
            <FormattedMessage
              id="ui-invoice.voucher.voucherLinesTitle.externalAccountNumber"
              values={{ externalAccountNumber }}
            />
          </Headline>

          <MultiColumnList
            columnMapping={columnMapping}
            columnWidths={columnWidths}
            contentData={lines}
            formatter={resultsFormatter}
            visibleColumns={visibleColumns}
          />

          <Row end="xs">
            <Headline>
              <FormattedMessage id="ui-invoice.voucher.voucherLines.total" />
              <AmountWithCurrencyField
                amount={getTotalAmount(lines)}
                currency={currency}
              />
            </Headline>
          </Row>
        </Fragment>
      );
    })
  );
};

VoucherLinesDetails.propTypes = {
  voucherLines: PropTypes.arrayOf(PropTypes.object).isRequired,
  currency: PropTypes.string.isRequired,
};

export default VoucherLinesDetails;
