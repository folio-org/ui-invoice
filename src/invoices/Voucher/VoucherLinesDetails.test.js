import { render } from '@folio/jest-config-stripes/testing-library/react';

import { batchVoucherLine } from '../../../test/jest/fixtures';

import VoucherLinesDetails from './VoucherLinesDetails';

const defaultProps = {
  voucherLines: [batchVoucherLine],
  currency: 'USD',
};
const renderVoucherLinesDetails = (props = defaultProps) => render(
  <VoucherLinesDetails {...props} />,
);

describe('VoucherLinesDetails', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', async () => {
    const { asFragment } = renderVoucherLinesDetails();

    expect(asFragment()).toMatchSnapshot();
  });
});
