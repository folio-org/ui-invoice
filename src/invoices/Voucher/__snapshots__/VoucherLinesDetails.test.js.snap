// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoucherLinesDetails should render correct structure 1`] = `
<DocumentFragment>
  <div
    class="headline size-medium margin-medium font-weight-bold"
    data-test-headline="true"
  >
    ui-invoice.voucher.voucherLinesTitle.externalAccountNumber
  </div>
  <div
    class="mclContainer"
    style="position: relative; overflow: hidden; height: auto; width: 100%;"
    tabindex="-1"
  >
    <div
      aria-live="assertive"
      aria-relevant="additions"
      class="sr-only"
    />
    <div
      aria-rowcount="2"
      class=""
      data-total-count="0"
      role="grid"
      tabindex="0"
    >
      <div
        class="mclHeaderContainer"
      >
        <div
          aria-rowindex="1"
          class="mclHeaderRow"
          role="row"
          style="display: flex; width: 17px;"
        >
          <div
            aria-sort="none"
            class="mclHeader"
            id="list-column-linenumber"
            role="columnheader"
            style="flex: 0 0 auto;"
          >
            ui-invoice.voucher.voucherLines.lineNumber
          </div>
          <div
            aria-sort="none"
            class="mclHeader"
            id="list-column-group"
            role="columnheader"
            style="flex: 0 0 auto;"
          >
            ui-invoice.voucher.voucherLines.group
          </div>
          <div
            aria-sort="none"
            class="mclHeader"
            id="list-column-fundcode"
            role="columnheader"
            style="flex: 0 0 auto;"
          >
            ui-invoice.voucher.voucherLines.fundCode
          </div>
          <div
            aria-sort="none"
            class="mclHeader"
            id="list-column-externalaccountnumber"
            role="columnheader"
            style="flex: 0 0 auto;"
          >
            ui-invoice.voucher.voucherLines.externalAccountNumber
          </div>
          <div
            aria-sort="none"
            class="mclHeader"
            id="list-column-amount"
            role="columnheader"
            style="flex: 0 0 auto;"
          >
            ui-invoice.voucher.voucherLines.amount
          </div>
        </div>
      </div>
      <div
        class="mclScrollable"
        style="position: static; width: 100%;"
        tabindex="0"
      >
        <div
          class="mclRowContainer"
          role="rowgroup"
          style="position: static;"
        >
          <div
            aria-rowindex="2"
            class="mclRowFormatterContainer"
            data-row-index="row-0"
            role="row"
            style="position: static;"
          >
            <div
              class="mclRow mclIsOdd mclIsInteractive"
              data-row-inner="0"
              style="min-width: 0px;"
            >
              <div
                class="mclCell mclCellStyle"
                role="gridcell"
              >
                1
              </div>
              <div
                class="mclCell mclCellStyle"
                role="gridcell"
              >
                <span>
                  -
                </span>
              </div>
              <div
                class="mclCell mclCellStyle"
                role="gridcell"
              >
                , , 
              </div>
              <div
                class="mclCell mclCellStyle"
                role="gridcell"
              >
                54321098
              </div>
              <div
                class="mclCell mclCellStyle"
                role="gridcell"
              >
                $23.45
              </div>
            </div>
          </div>
          <div
            class="mclEndOfListContainer"
            data-end-of-list="1"
          >
            <div
              class="mclCenteredContainer"
              style="visibility: hidden; height: 0px; padding: 0px; width: 0px;"
            >
              <div
                class="textCentered"
                data-test-layout="true"
              >
                <div
                  class="mclEndOfList"
                  data-test-layout="true"
                >
                  <span>
                    Icon
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row end-xs"
  >
    <div
      class="headline size-medium margin-medium font-weight-bold"
      data-test-headline="true"
    >
      ui-invoice.voucher.voucherLines.total$23.45
    </div>
  </div>
</DocumentFragment>
`;
